### PR TITLE
Br54 test module for batch

### PR DESCRIPTION
**Pull request by zengxyu**

### CODE REVIEW

File: pytest/blueprints/constructables/test_higher_order_iterator_constructables.py [hunk 1]
```diff
@@ -36,6 +50,32 @@ def informed_iterators(self, repository_requirement):
...
+    def collator_type(self):
+        class MNISTCollator(Collator):
+            def __call__(self, batch: List[torch.Tensor]):
+                pass
+
+        return MNISTCollator
+
+    @pytest.fixture
+    def data_collator(self, informed_iterators, collator_type):
+        requirements = {"iterators": Requirement(components=informed_iterators, subscription=["train", "test"])}
```
Comment (le1nux) [inline pytest/blueprints/constructables/test_higher_order_iterator_constructables.py:63]: Collator does not have a dependency for iterators. A dataloader does.

File: pytest/blueprints/constructables/test_higher_order_iterator_constructables.py [hunk 2]
```diff
@@ -152,6 +236,109 @@ def test_constructable(self, informed_iterators):
...
+#         iterators = constructable.construct()
+#         iterator_train_encoded = iterators["train"]
+#         sample, target, tag = iterator_train_encoded[0]
+#         assert list(sample.shape) == [28, 28]
+#         assert isinstance(target, int)
+#
+#         assert isinstance(iterator_train_encoded, InformedDatasetIteratorIF)
+
+
+class TestDataCollatorConstructable(IteratorFixtures):
```
Comment (le1nux) [inline pytest/blueprints/constructables/test_higher_order_iterator_constructables.py:256]: this is not a higher level iterator constructable. I think we should move this.

File: pytest/blueprints/constructables/test_higher_order_iterator_constructables.py [hunk 3]
```diff
@@ -152,6 +236,109 @@ def test_constructable(self, informed_iterators):
...
+        collator_type = collator_type
+        constructable = DataCollatorConstructable(component_identifier="one_hot_targets_component",
+                                                  requirements=requirements,
+                                                  collator_params=collator_params,
+                                                  collator_type=collator_type)
+        iterators = constructable.construct()
+        assert isinstance(iterators, collator_type)
+
+
+class TestDataLoadersConstructable(IteratorFixtures):
```
Comment (le1nux) [inline pytest/blueprints/constructables/test_higher_order_iterator_constructables.py:269]: this is not a higher level iterator constructable. I think we should move this.

File: pytest/blueprints/constructables/test_higher_order_iterator_constructables.py [hunk 4]
```diff
@@ -152,6 +236,109 @@ def test_constructable(self, informed_iterators):
...
+                                                 batch_size=batch_size,
+                                                 sampling_strategies=sampling_strategies,
+                                                 drop_last=drop_last
+                                                 )
+        iterators = constructable.construct()
+        assert isinstance(iterators["train"].sampler, RandomSampler)
+        assert isinstance(iterators["test"].sampler, WeightedRandomSampler)
+
+
+class TestOptimizerConstructable:
```
Comment (le1nux) [inline pytest/blueprints/constructables/test_higher_order_iterator_constructables.py:288]: this is not a higher level iterator constructable. I think we should move this.

Comment (le1nux) [reply pytest/blueprints/constructables/test_higher_order_iterator_constructables.py:288]: same goes for all the constructables below

File: pytest/blueprints/constructables/test_higher_order_iterator_constructables.py [hunk 5]
```diff
@@ -81,6 +122,49 @@ def test_constructable(self, informed_iterators):
         # print(DatasetIteratorReportGenerator.generate_report(iterator_full))
 
 
+class TestInMemoryDatasetIteratorConstructable(IteratorFixtures):
+    def test_constructable(self, informed_iterators):
```
Comment (le1nux) [inline pytest/blueprints/constructables/test_higher_order_iterator_constructables.py:126]: Ok, so now we test if the component can be **constructed**. I think in a second step we should also test the **functionality** of the constructed component as well. In the case of the inmemory dataset iterator, we would check if everything is loaded into memory correctly, for instance. We should discuss this. :)

File: pytest/blueprints/constructables/test_higher_order_iterator_constructables.py [hunk 6]
```diff
@@ -81,6 +122,49 @@ def test_constructable(self, informed_iterators):
...
+        shuffled_sample, shuffled_target, shuffled_tag = shuffled_iterators["train"][0]
+
+        assert list(sample.shape) == [28, 28]
+        assert isinstance(target, int)
+
+        assert isinstance(shuffled_iterators["train"], InformedDatasetIterator)
+        assert isinstance(shuffled_iterators["test"], InformedDatasetIterator)
+
+        # assert if shuffled_sample != sample
+        assert (shuffled_sample != sample).any()
```
Comment (le1nux) [inline pytest/blueprints/constructables/test_higher_order_iterator_constructables.py:165]: Even if we shuffle, couldn't it also be the that just by chance the two match?

Comment (le1nux) [reply pytest/blueprints/constructables/test_higher_order_iterator_constructables.py:165]: (even if unlikely)

File: pytest/blueprints/constructables/test_higher_order_iterator_constructables.py [hunk 7]
```diff
@@ -152,6 +236,109 @@ def test_constructable(self, informed_iterators):
         # print(DatasetIteratorReportGenerator.generate_report(iterator_train_encoded))
 
 
+# class TestOneHotEncodedTargetsIteratorConstructable(IteratorFixtures):
```
Comment (le1nux) [inline pytest/blueprints/constructables/test_higher_order_iterator_constructables.py:239]: can this be removed?

File: pytest/blueprints/constructables/test_higher_order_iterator_constructables.py [hunk 8]
```diff
@@ -152,6 +236,109 @@ def test_constructable(self, informed_iterators):
...
+class TestDataCollatorConstructable(IteratorFixtures):
+    def test_constructable(self, informed_iterators, collator_type):
+        requirements = {"iterators": Requirement(components=informed_iterators, subscription=["train", "test"])}
+        collator_params = {}
+        collator_type = collator_type
+        constructable = DataCollatorConstructable(component_identifier="one_hot_targets_component",
+                                                  requirements=requirements,
+                                                  collator_params=collator_params,
+                                                  collator_type=collator_type)
+        iterators = constructable.construct()
```
Comment (le1nux) [inline pytest/blueprints/constructables/test_higher_order_iterator_constructables.py:265]: iterators name is misleading

File: pytest/blueprints/constructables/test_higher_order_iterator_constructables.py [hunk 9]
```diff
@@ -152,6 +236,109 @@ def test_constructable(self, informed_iterators):
...
+#         sample, target, tag = iterator_train_encoded[0]
+#         assert list(sample.shape) == [28, 28]
+#         assert isinstance(target, int)
+#
+#         assert isinstance(iterator_train_encoded, InformedDatasetIteratorIF)
+
+
+class TestDataCollatorConstructable(IteratorFixtures):
+    def test_constructable(self, informed_iterators, collator_type):
+        requirements = {"iterators": Requirement(components=informed_iterators, subscription=["train", "test"])}
```
Comment (le1nux) [inline pytest/blueprints/constructables/test_higher_order_iterator_constructables.py:258]: collator does not have this requirement

File: pytest/blueprints/constructables/test_higher_order_iterator_constructables.py [hunk 10]
```diff
@@ -152,6 +236,109 @@ def test_constructable(self, informed_iterators):
...
+        iterators = constructable.construct()
+        assert isinstance(iterators, collator_type)
+
+
+class TestDataLoadersConstructable(IteratorFixtures):
+    def test_constructable(self, informed_iterators, data_collator):
+        requirements = {"iterators": Requirement(components=informed_iterators, subscription=["train", "test"]),
+                        "data_collator": Requirement(components=data_collator, subscription=["train", "test"])}
+        batch_size: int = 16
+        sampling_strategies = {'train': {"strategy": "RANDOM", "seed": 0},
```
Comment (le1nux) [inline pytest/blueprints/constructables/test_higher_order_iterator_constructables.py:274]: we also have IN_ORDER now, could you add that as well please?

File: pytest/blueprints/constructables/test_higher_order_iterator_constructables.py [hunk 11]
```diff
@@ -152,6 +236,109 @@ def test_constructable(self, informed_iterators):
...
+
+    def test_constructable(self):
+        optimizer_key = "SGD"
+        params = {"lr": 1.0, "momentum": 0.9}
+
+        constructable = OptimizerConstructable(optimizer_key=optimizer_key, params=params)
+        iterators = constructable.construct()
+
+        assert isinstance(iterators, OptimizerAdapter)
+        assert iterators._optimizer_class == SGD
```
Comment (le1nux) [inline pytest/blueprints/constructables/test_higher_order_iterator_constructables.py:298]: Not an iterator. Same hold for the variables below

File: pytest/blueprints/constructables/test_higher_order_iterator_constructables.py [hunk 12]
```diff
@@ -152,6 +236,109 @@ def test_constructable(self, informed_iterators):
...
+        iterators = constructable.construct()
+        assert len(iterators._store) != 0
+
+
+class TestPredictionPostProcessingRegistryConstructable:
+    def test_constructable(self):
+        constructable = PredictionPostProcessingRegistryConstructable()
+        iterators = constructable.construct()
+        assert len(iterators._store) != 0
+        assert iterators[
```
Comment (le1nux) [inline pytest/blueprints/constructables/test_higher_order_iterator_constructables.py:328]: I guess these checks are a bit redundant. They don't test functionality but rather "static assignments"

**Review by le1nux — CHANGES_REQUESTED** [review]
Thanks for the PR. I left a couple of remarks which we need to address before merging it :-)

File: pytest/blueprints/constructables/test_higher_order_iterator_constructables.py [hunk 13]
```diff
@@ -36,6 +50,32 @@ def informed_iterators(self, repository_requirement):
         iterators = constructable.construct()
         return iterators
 
+    @pytest.fixture
+    def collator_type(self):
+        class MNISTCollator(Collator):
+            def __call__(self, batch: List[torch.Tensor]):
+                pass
+
+        return MNISTCollator
```
[the remark this answers is not in the document]
Comment (zengxyu) [reply pytest/blueprints/constructables/test_higher_order_iterator_constructables.py:59]: you are right, I will rename it

**Review by le1nux — CHANGES_REQUESTED** [review]
I think we are almost there. Just a few tweaks here and there and we can merge :-)

**Review by le1nux — APPROVED** [review]
Looks good to me! Good work! :-)

**Pull request merged**